### PR TITLE
core: use loop.Keystore, support arbitrarily prefixed Cosmos addresses

### PR DESCRIPTION
**Pull request by cfal**
BCI-1740

- adds a loop.Keystore implementation using keystore.Cosmos and uses it; note that this is still for unprefixed keys, such that key management using the key controller follows existing conventions (ie. manage a single Cosmos keystore key used across Cosmos chains)
- adds a cosmostxm.KeystoreAdapter that translates loop.Keystore accounts into prefixed Cosmos addresses
- updates cosmostxm.KeyWrapper to wrap a KeystoreAdapter instance for a specific prefixed Cosmos address

### CODE REVIEW

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+type accountInfo struct {
+	Account string
+	PubKey  *secp256k1.PubKey
+}
+
+// An adapter for a Cosmos loop.Keystore to translate public keys into bech32-prefixed account addresses.
+type KeystoreAdapter struct {
+	keystore        loop.Keystore
+	accountPrefix   string
+	addressToPubKey map[string]*accountInfo
```
Comment (krehermann) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:27]: does this need a mutex? i see it being mutated in update and lookup

**Review by krehermann — COMMENTED** [review]
looks good to me upto the question about a mutex.

out of curiousity - how burdensome was it to understand what to write? i'm excited that you could figure it out and i'm wondering if we could have made it easier to understand. based on the code, i'm guessing that starknet was a good example

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+func (ka *KeystoreAdapter) Accounts(ctx context.Context) ([]string, error) {
+	if err := ka.updateMapping(); err != nil {
+		return nil, err
+	}
+	accounts := make([]string, 0, len(ka.addressToPubKey))
+	for k := range ka.addressToPubKey {
+		accounts = append(accounts, k)
+	}
+	return accounts, nil
```
Comment (jmank88) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:122]:
```suggestion
	accounts := make([]string, 0, len(ka.addressToPubKey))
	returns maps.Keys(ka.addressToPubKey), nil
```

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+			PubKey:  &secp256k1.PubKey{Key: pubKeyBytes},
+		}
+	}
+
+	ka.addressToPubKey = addressToPubKey
+	return nil
+}
+
+func (ka *KeystoreAdapter) lookup(id string) (*accountInfo, error) {
+	ai, ok := ka.addressToPubKey[id]
```
Comment (krehermann) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:98]: this should be read protected with the mutex

File: core/chains/cosmos/cosmostxm/keystore_adapter.go (hunk 1 above)
Comment (krehermann) [reply core/chains/cosmos/cosmostxm/keystore_adapter.go:29]: nit/sytle: move the mutex definition above the resource that it protects

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			return err
+		}
+
+		if len(pubKeyBytes) != secp256k1.PubKeySize {
+			return errors.New("length of pubkey is incorrect")
+		}
+
+		sha := sha256.Sum256(pubKeyBytes)
+		hasherRIPEMD160 := ripemd160.New()
+		hasherRIPEMD160.Write(sha[:]) // does not error
```
Comment (jmank88) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:76]: Another option: :shrug: 
```suggestion
		_, _ = hasherRIPEMD160.Write(sha[:])
```

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+
+	ka.addressToPubKey = addressToPubKey
+	return nil
+}
+
+func (ka *KeystoreAdapter) lookup(id string) (*accountInfo, error) {
+	ka.mutex.Lock()
+	defer ka.mutex.Unlock()
```
Comment (jmank88) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:96]: We might want to add an optimized read lock path here. Or maybe just switch to a `sync.Map` for simplicity?

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+			PubKey:  &secp256k1.PubKey{Key: pubKeyBytes},
+		}
+	}
+
+	ka.addressToPubKey = addressToPubKey
+	return nil
+}
+
+func (ka *KeystoreAdapter) lookup(id string) (*accountInfo, error) {
+	ka.mutex.Lock()
```
Comment (calvwang9) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:95]: for my own understanding, is there a reason to put the mutex lock/unlock in the internal functions vs top level public functions (Sign, PubKey)?

File: core/chains/cosmos/cosmostxm/keystore_adapter.go (hunk 5 above)
Comment (krehermann) [reply core/chains/cosmos/cosmostxm/keystore_adapter.go:96]: yeah, i would opt for RW mutex for now. i meant to convey that with my comment about read protection, but i wasn't clear at all.

File: core/chains/cosmos/cosmostxm/keystore_adapter.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	ai, ok := ka.addressToPubKey[id]
+	ka.mutex.RUnlock()
+	if !ok {
+		// try updating the mapping once, incase there was an update on the keystore.
+		if err := ka.updateMapping(); err != nil {
+			return nil, err
+		}
+		ka.mutex.RLock()
+		ai, ok = ka.addressToPubKey[id]
+		ka.mutex.RUnlock()
```
Comment (jmank88) [inline core/chains/cosmos/cosmostxm/keystore_adapter.go:107]: nit/ you can drop this additional read lock if you grab this info while still holding the write lock. Maybe moving the lock out of updateMapping would be simplest.

File: core/chains/cosmos/cosmostxm/helpers_test.go [hunk 8]
```diff
@@ -28,3 +29,16 @@ func (txm *Txm) MarshalMsg(msg sdk.Msg) (string, []byte, error) {
...
+
+func (ka *KeystoreAdapter) Accounts(ctx context.Context) ([]string, error) {
+	ka.mutex.Lock()
+	err := ka.updateMappingLocked()
+	if err != nil {
+		ka.mutex.Unlock()
+		return nil, err
+	}
+	addresses := maps.Keys(ka.addressToPubKey)
+	ka.mutex.Unlock()
```
Comment (jmank88) [inline core/chains/cosmos/cosmostxm/helpers_test.go:41]: nit/ could defer unlock here for simplicity:
```suggestion
	defer ka.mutex.Unlock()
	err := ka.updateMappingLocked()
	if err != nil {
		return nil, err
	}
	addresses := maps.Keys(ka.addressToPubKey)
```

**Review by jmank88 — APPROVED** [review]

File: core/chains/cosmos/cosmostxm/helpers_test.go (hunk 8 above)
Comment (krehermann) [reply core/chains/cosmos/cosmostxm/helpers_test.go:41]: +1

**Review by krehermann — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Comment by augustbleeds** [comment]
^ keeping the branch restored so it can be used for tests on chainlink-cosmos

**Pull request merged**